src/internal/plugins/controllers/alert-flash-content.ts

```diff
@@ -0,0 +1,104 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import debounce from '../../debounce';
+
+// this code should not depend on React typings, because it is portable between major versions
+interface RefShim<T> {
+  current: T | null;
+}
+
+export interface AlertFlashContentContext {
+  type: string;
+  headerRef: RefShim<HTMLElement>;
+  contentRef: RefShim<HTMLElement>;
+}
+
+export type ReplacementType = 'original' | 'remove' | 'replaced';
+
+export interface ReplacementApi {
+  hideHeader(): void;
+  restoreHeader(): void;
+  replaceHeader(replacer: (container: HTMLElement) => void): void;
+  hideContent(): void;
+  restoreContent(): void;
+  replaceContent(replacer: (container: HTMLElement) => void): void;
+}
+
+export interface AlertFlashContentResult {
+  update: () => void;
+  unmount: (containers: { replacementHeaderContainer: HTMLElement; replacementContentContainer: HTMLElement }) => void;
+}
+
+export interface AlertFlashContentConfig {
+  id: string;
+  runReplacer: (context: AlertFlashContentContext, replacementApi: ReplacementApi) => AlertFlashContentResult;
+}
+
+export type AlertFlashContentRegistrationListener = (provider?: AlertFlashContentConfig) => void | (() => void);
```

Comment:
Why is this `(() => void)` part needed?

Reply:
oh, get it now, it is a return value `void | (() => void)`.

I think for more convenience we could always return a cleanup callback, even if noop

```suggestion
export type AlertFlashContentRegistrationListener = (provider?: AlertFlashContentConfig) => () => void;
```

Reply:
Updated